File: linked_list/LL_heap_pq/LL_maxheap_pq.py
class PriorityQueue:
    """
    PriorityQueue: 정수를 저장하고 관리하는 최대 힙 구현.
    """
    def __init__(self):
        """
        PriorityQueue 초기화.
        내부적으로 요소를 저장할 빈 리스트를 생성합니다.
        """
        self.elements = []

    def bubble_up(self, index):
        """
        새로운 요소를 삽입한 후 최대 힙 속성을 유지하기 위해 부모 노드와 비교하여 위로 이동.
        - 삽입 시 호출됩니다.
        """
        while index > 0 and self.elements[index] > self.elements[(index - 1) // 2]:
            self.elements[index], self.elements[(index - 1) // 2] = self.elements[(index - 1) // 2], self.elements[index]
            index = (index - 1) // 2

    def bubble_down(self, index):
        """
        루트에서 요소를 제거한 후 최대 힙 속성을 유지하기 위해 자식 노드와 비교하며 아래로 이동.
        - 제거 시 호출됩니다.
        """
        size = len(self.elements)
        while index * 2 + 1 < size:
            child = index * 2 + 1
            if child + 1 < size and self.elements[child + 1] > self.elements[child]:
                child += 1
            if self.elements[index] >= self.elements[child]:
                break
            self.elements[index], self.elements[child] = self.elements[child], self.elements[index]
            index = child

    def max_heapify(self, size, index):
        """
        특정 노드에서 시작하여 서브트리를 최대 힙으로 복구.
        - 힙 정렬에서 사용됩니다.
        """
        largest = index
        left = 2 * index + 1
        right = 2 * index + 2

        if left < size and self.elements[left] > self.elements[largest]:
            largest = left
        if right < size and self.elements[right] > self.elements[largest]:
            largest = right

        if largest != index:
            self.elements[index], self.elements[largest] = self.elements[largest], self.elements[index]
            self.max_heapify(size, largest)

    def sort(self):
        """
        우선순위 큐의 요소를 정렬.
        - max_heapify를 사용하여 힙 정렬 방식으로 정렬합니다.
        """
        original_size = len(self.elements)

        # 최대 힙 생성
        for i in range(len(self.elements) // 2 - 1, -1, -1):
            self.max_heapify(len(self.elements), i)

        # 힙 요소를 하나씩 정렬
        for i in range(len(self.elements) - 1, 0, -1):
            self.elements[0], self.elements[i] = self.elements[i], self.elements[0]
            self.max_heapify(i, 0)

        # 원래 크기 복원 (히프 속성을 유지하며)
        self.elements = self.elements[:original_size]

    def insert(self, value):
        """
        새로운 정수 값을 우선순위 큐에 삽입.
        - bubble_up을 호출하여 힙 속성을 유지합니다.
        """
        self.elements.append(value)
        self.bubble_up(len(self.elements) - 1)

    def pop(self):
        """
        우선순위 큐의 최대값(루트)을 제거하고 반환.
        - bubble_down을 호출하여 힙 속성을 유지합니다.
        """
        if not self.elements:
            raise IndexError("우선순위 큐가 비어 있습니다!")
        max_value = self.elements[0]
        self.elements[0] = self.elements.pop()
        self.bubble_down(0)
        return max_value

    def print(self):
        """
        현재 우선순위 큐의 모든 요소를 출력.
        """
        print(self.elements)


class Node:
    """
    연결 리스트의 노드 클래스.
    각 노드는 PriorityQueue(우선순위 큐)를 포함하며, 다음 노드를 가리킵니다.
    """
    def __init__(self):
        """
        Node 초기화: 빈 PriorityQueue와 다음 노드 포인터를 None으로 설정.
        """
        self.heap = PriorityQueue()
        self.next = None


class LinkedList:
    """
    Node로 구성된 단일 연결 리스트.
    """
    def __init__(self):
        """
        LinkedList 초기화: head를 None으로 설정.
        """
        self.head = None

    def add_node(self):
        """
        새 노드를 연결 리스트의 맨 앞에 추가.
        """
        new_node = Node()
        new_node.next = self.head
        self.head = new_node

    def print_heaps(self):
        """
        연결 리스트의 각 노드에 포함된 힙을 출력.
        """
        current = self.head
        index = 0
        while current:
            print(f"노드 {index}의 힙: ", end="")
            current.heap.print()
            current = current.next
            index += 1


# 테스트 코드
print("정수형 힙을 포함한 단일 연결 리스트:")

ll = LinkedList()
ll.add_node()
ll.add_node()

# 첫 번째 노드의 힙에 요소 추가
ll.head.heap.insert(10)
ll.head.heap.insert(20)
ll.head.heap.insert(5)

# 두 번째 노드의 힙에 요소 추가
ll.head.next.heap.insert(15)
ll.head.next.heap.insert(25)
ll.head.next.heap.insert(30)

print("삽입 후 연결 리스트의 힙:")
ll.print_heaps()

# 첫 번째 힙 정렬
print("\n첫 번째 힙 정렬 후:")
ll.head.heap.sort()
ll.head.heap.print()

File: linked_list/LL_heap_pq/LL_maxheap_pq.cpp
#include <iostream>
#include <vector>
using namespace std;

/*
 * PriorityQueue 클래스: 정수 요소를 저장하며 최대 힙(Max-Heap) 구조를 유지.
 * Node 클래스: PriorityQueue를 포함하며, 다음 노드를 가리키는 포인터를 포함.
 * LinkedList 클래스: Node 객체들로 구성된 단일 연결 리스트를 관리.
 *
 * 주요 기능:
 * - PriorityQueue: 정수를 기반으로 삽입, 삭제, 출력, 힙 정렬 지원.
 * - Node: PriorityQueue와 다음 노드 포인터를 포함.
 * - LinkedList: Node를 관리하며, 힙 데이터 출력 기능 제공.
 */

// PriorityQueue 클래스 정의
class PriorityQueue {
private:
    vector<int> elements; // 내부 벡터: 정수 요소를 저장

    // 힙 삽입 시 최대 힙 속성을 유지하기 위한 버블 업 연산
    void bubbleUp(int index) {
        while (index > 0 && elements[index] > elements[(index - 1) / 2]) {
            swap(elements[index], elements[(index - 1) / 2]);
            index = (index - 1) / 2;
        }
    }

    // 힙 삭제 시 최대 힙 속성을 유지하기 위한 버블 다운 연산
    void bubbleDown(int index) {
        int size = elements.size();
        while (index * 2 + 1 < size) {
            int child = index * 2 + 1;
            if (child + 1 < size && elements[child + 1] > elements[child]) {
                child++;
            }
            if (elements[index] >= elements[child]) {
                break;
            }
            swap(elements[index], elements[child]);
            index = child;
        }
    }

    // 힙 정렬을 위한 max-heapify 연산
    void maxHeapify(int size, int index) {
        int largest = index;
        int left = 2 * index + 1;
        int right = 2 * index + 2;

        if (left < size && elements[left] > elements[largest]) {
            largest = left;
        }
        if (right < size && elements[right] > elements[largest]) {
            largest = right;
        }

        if (largest != index) {
            swap(elements[index], elements[largest]);
            maxHeapify(size, largest);
        }
    }

public:
    // PriorityQueue에 정수 값을 삽입
    void insert(int value) {
        elements.push_back(value);
        bubbleUp(elements.size() - 1);
    }

    // PriorityQueue에서 최대값을 제거하고 반환
    int pop() {
        if (elements.empty()) {
            throw runtime_error("우선순위 큐가 비어 있습니다!");
        }
        int maxValue = elements[0];
        elements[0] = elements.back();
        elements.pop_back();
        bubbleDown(0);
        return maxValue;
    }

    // PriorityQueue의 모든 요소를 출력
    void print() const {
        for (int value : elements) {
            cout << value << " ";
        }
        cout << endl;
    }

    // PriorityQueue를 힙 정렬
    void sort() {
        int originalSize = elements.size();

        // 최대 힙을 생성
        for (int i = elements.size() / 2 - 1; i >= 0; i--) {
            maxHeapify(elements.size(), i);
        }

        // 힙 요소를 하나씩 제거하며 정렬
        for (int i = elements.size() - 1; i > 0; i--) {
            swap(elements[0], elements[i]); // 최대값을 끝으로 이동
            maxHeapify(i, 0); // 축소된 힙에서 힙 속성 복구
        }

        elements.resize(originalSize); // 원래 크기로 복원
    }

    // PriorityQueue가 비었는지 확인
    bool isEmpty() const {
        return elements.empty();
    }
};

// Node 클래스 정의
class Node {
public:
    PriorityQueue heap; // 이 노드의 PriorityQueue
    Node* next;         // 다음 노드를 가리키는 포인터

    // Node 생성자: 다음 노드 포인터를 nullptr로 초기화
    Node() : next(nullptr) {}
};

// LinkedList 클래스 정의
class LinkedList {
private:
    Node* head; // 리스트의 첫 번째 노드

public:
    // LinkedList 생성자: head를 nullptr로 초기화
    LinkedList() : head(nullptr) {}

    // 새로운 노드를 리스트 앞에 추가
    void addNode() {
        Node* newNode = new Node();
        newNode->next = head;
        head = newNode;
    }

    // 리스트의 첫 번째 노드 반환
    Node* getHead() const {
        return head;
    }

    // 리스트의 각 노드에 포함된 힙 출력
    void printHeaps() const {
        Node* current = head;
        int index = 0;
        while (current != nullptr) {
            cout << "노드 " << index++ << "의 힙: ";
            current->heap.print();
            current = current->next;
        }
    }

    // LinkedList 소멸자: 모든 노드 삭제
    ~LinkedList() {
        Node* current = head;
        while (current != nullptr) {
            Node* temp = current;
            current = current->next;
            delete temp;
        }
    }
};

// 메인 함수: 구현 테스트
int main() {
    cout << "정수형 힙을 포함한 단일 연결 리스트:\n";

    LinkedList list;
    list.addNode();
    list.addNode();

    Node* head = list.getHead();

    // 첫 번째 노드의 힙에 요소 삽입
    head->heap.insert(15);
    head->heap.insert(20);
    head->heap.insert(10);

    // 두 번째 노드의 힙에 요소 삽입
    head->next->heap.insert(25);
    head->next->heap.insert(30);
    head->next->heap.insert(5);

    cout << "삽입 후 연결 리스트의 힙:\n";
    list.printHeaps();

    // 첫 번째 노드의 힙 정렬
    cout << "\n첫 번째 힙 정렬 후:\n";
    head->heap.sort();
    head->heap.print();

    return 0;
}